### PR TITLE
813 - Persist portal metadata computed file and upload after creation

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number

Closing #813

Feature branch: `feature/portal-metadata-command`

## Purpose/Implementation Notes
The following methods were added to the `create_portal_metadata` management command:
- Django built-in `save` method to persist `computed_file` in the `ComputedFile` model's table
- `ComputedFile::upload_s3_file` to upload `portal_metadata.zip` to the S3 bucket 

Adjusted the test:
- Added the assertion for the computed file to `test_create_portal_metadata`

## Types of changes
- New feature (non-breaking change which adds functionality)

## Functional tests

N/A

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

N/A

### CODE REVIEW

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:17]:
```suggestion
```

Don't need to do this, we can reference the same config settings for now

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:27]:
```suggestion
            "--clean-up-output-data", action=BooleanOptionalAction, default=settings.PRODUCTION
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:29]:
```suggestion
        parser.add_argument("--update-s3", action=BooleanOptionalAction, default=settings.UPDATE_S3_DATA)
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:36]:
```suggestion
        update_s3 = kwargs.get("update_s3", settings.UPDATE_S3_DATA)
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:47]:
```suggestion
            if update_s3:
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:49]: This is now out of date. We need to pull in the `scpca_portal.s3` module and call `s3.upload_output_file(computed_file.s3_key, bucket_name=computed_file.s3_bucket)`

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:33]: We will want to match the existing behavior around testing readmes once we merge in #839

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: Replace the method to a new one from the `s3` module  at ffcc45ee90a7635d626843cd4117922b51421a0f.

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: Changed the verb from `upload` to `update` at ffcc45ee90a7635d626843cd4117922b51421a0f.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 0c7b5ae416c940b666543f3e32728b3eae7a13f8.

Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: Removed the constant `ENCODING` at 0c7b5ae416c940b666543f3e32728b3eae7a13f8  and added the `TODO` comment indicating that after PR #839 is merged into `dev`, readme file format testing will be added (this `assertProjectReadmeContains` method will be removed).

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:29]: This is the line to prevent tests from running in parallel. 

Additionally, I've temporarily disabled `isort` from modifying the import order of the management commands. As mentioned in the eng meeting, the import for `create_portal_metadata` must come before `configure_aws_cli` to avoid this specific issue. 

I was thinking of removing the duplicate logging messages to clean up the logs, but I’m not sure if that’s the best move. Please let me know if you have any insights or feedback on this, thank you David!

Comment (davidsmejia) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py:29]: This can just be removed. As per our conversation

Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: This has been removed at baa659c0c1a8ec117f7506944ad679a748df2ac7, thank you for the insights, David!

**Comment by nozomione** [comment]
I've applied your feedback. Also please see the additional update in https://github.com/AlexsLemonade/scpca-portal/pull/825#discussion_r1725756483. This PR is now ready for another review, thank you David!

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:42]:
```suggestion
                s3.upload_output_file(computed_file.s3_key, computed_file.s3_bucket)
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:40]:
```suggestion
            if update_s3:
```
lets break up kwarg and add typehints to the function signature too.
```python
def create_portal_metadata(clean_up_output_data: bool, update_s3: bool, **kwargs):
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 609dc068afede2c66c6a427d0435792e7661a2fc.

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py]: See at 609dc068afede2c66c6a427d0435792e7661a2fc.

**Comment by nozomione** [comment]
I've applied your feedback and this PR is ready for your review. Thank you David!

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:76]: I think here we should be calling the same assertions independent of the command running successfully. That is what we are here to test / ensure.

I rewrote the test below but I think for better coverage we should also assert that all of the 
`GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG` values have been applied to the computed file as well.

```suggestion
        self.processor.create_portal_metadata(
            clean_up_output_data=False, update_s3=True
        )

        computed_files = ComputedFile.objects.filter(portal_metadata_only=True)
        self.assertEqual(portal_computed_files.count(), 1)
        
        computed_file = portal_computed_files.first()
        self.assertEqual(computed_file.size_in_bytes, 8469)
        mock_upload_output_file.assert_called_once_with(
            computed_file.s3_key, computed_file.s3_bucket
        )
```

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/management/commands/create_portal_metadata.py:46]: I've gone ahead and filed the issue here #869 👍!

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: Thank you for the rewrite, it looks great! I've also applied your suggestion to assert that all field values match the download config file at d1f8a256a9fe6c9edda6e79b438f354a34e174f9 👍!

**Comment by nozomione** [comment]
I've applied your feedback and this PR is ready for another look. Thank you David!

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:72]:
```suggestion
        self.assertEqualWithVariance(computed_file.size_in_bytes, 8460, 50)
        
        # outside of this test
        def assertEqualWithVariance(self, value, expected, variance):
            self.assertGreater(expected + variance, value)
            self.assertLess(expected - variance, value)
        
```

If you find that the value changes and you cannot reliably assert that the value matches you can create a test that allows for some wiggle room.

**Review by davidsmejia — APPROVED** [review]
LGTM if you find that testing the value does not consistently pass, I left a comment with an alternative.

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:36]: You can use the walrus operator here.

File: api/scpca_portal/management/commands/create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/management/commands/create_portal_metadata.py:42]: Let's move updating above saving it to the DB, so that if it fails on upload we exit and the file is unavailable.

**Review by davidsmejia — COMMENTED** [review]
Sorry one more comment about ordering so that we don't end up with an object in the DB that no one can download.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: This is such a great alternative! I've applied it at daccc79a58a3a61df45f50277bc18a8368e21432 👍 

I’ve also added a new method called `assertFields`, which asserts that the fields in the portal metadata computed file match the download configuration settings. We talked about this optimization idea during our 1:1, so I went ahead and applied the input as well!

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: I've applied this feedback at 75ed5b3bbdba1d67e683f3de35e8d22b69a6819f👍

**Pull request merged**
